### PR TITLE
feat: Add GitHub Action for Dockerfile release update and slack notification

### PR DESCRIPTION
**Pull request by mamu0**
#### List of Changes

1. Adding the GitHub Action.
2. Added the `release.json` file, which contain the values of the latest release. This variables are used to store the tag and SHA of the latest installed version.
      ```
      {
        "current_tag": "vX.Y.Z",
        "current_sha": "XXXXXX"
      }
     ```
4. Making a minor fix to the .gitignore file.
5. Enabling “_Allow GitHub Actions to create and approve pull requests_” in `Repository Settings` -> `Actions` -> `General`.
6. Added a variable for SLACK WEBOOK SECRET to use inside the workflow

#### Motivation and Context

This PR aims to create a GitHub Action that checks if a new release has been published on the repository https://github.com/actions/runner. If a new release is found, the action retrieves the necessary information to update the FROM directive in the Dockerfile to reference the new release.

Additionally, a changelog has been added, which is an MD file that keeps track of the versions that have been installed over time.

#### How Has This Been Tested?

This has been tested by running the action multiple times to verify that the values were correct. The PR generation part has been tested in a private repository.

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: .github/workflows/check-release.yml
Comment (Krusty93) [inline .github/workflows/check-release.yml:25]:
```suggestion
        uses: actions/github-script@60a0d83039c74a4aee543508d2ffcb1c3799cdea # v7.0.1
```

File: .github/workflows/check-release.yml
Comment (Krusty93) [inline .github/workflows/check-release.yml:116]:
```suggestion
        uses: actions/github-script@60a0d83039c74a4aee543508d2ffcb1c3799cdea # v7.0.1
```

File: .github/workflows/check-release.yml
Comment (Krusty93) [inline .github/workflows/check-release.yml:147]: Not sure we can use this action

File: release.json
Comment (Krusty93) [inline release.json:1]: PR looks very good!
However, I'm not sure about this file. Would using repo vars more difficult?

Comment (mamu0) [reply release.json:1]: Thanks!
I tried to do it with gh commands, but it didn’t work as expected. However, I can try again in another way

File: .github/workflows/check-release.yml
Comment (mamu0) [reply .github/workflows/check-release.yml:147]: How can we verify it? In any case, we can convert it into a script step that executes git commands.

Comment (gunzip) [reply .github/workflows/check-release.yml:147]: I agree that it would be better to use `gh pr`

File: .github/workflows/check-release.yml
Comment (gunzip) [inline .github/workflows/check-release.yml:118]: it looks like this repo use semantic-release to update changelog using conventional commit mssages, may be we can reuse it someway here?

see https://github.com/semantic-release/release-notes-generator

File: release.json
Comment (Krusty93) [reply release.json:1]: You need a PAT with these permissions: https://github.com/pagopa/selfcare-onboarding/blob/ab774468aa2573399352d9430a1de788bb24fa13/.github/workflows/create_release_branch.yml#L91

We can use a bot user to create it

File: .github/workflows/check-release.yml
Comment (Krusty93) [inline .github/workflows/check-release.yml:96]: Env var is already set at higher scope job level

File: .github/workflows/check-release.yml
Comment (Krusty93) [inline .github/workflows/check-release.yml:159]: ditto

File: .github/workflows/check-release.yml
Comment (Krusty93) [inline .github/workflows/check-release.yml:166]: ditto

File: .github/workflows/check-release.yml
Comment (Krusty93) [inline .github/workflows/check-release.yml:202]: ditto

**Review by pasqualedevita — COMMENTED** [review]
Thank you @mamu0 this is an interesting contribution and improvements.

Anyway this points needs a discussion or review:

> 4. Enabling “Allow GitHub Actions to create and approve pull requests” in Repository Settings -> Actions -> General.

This is not possibile, in this way we grant a possibile not reviewed version. In other words is like use latest in dockerfile.

> 5. Added a variable for SLACK WEBOOK SECRET to use inside the workflow

Please check that all new secrets follow the technology standard to prevent leak or exposure.
https://pagopa.atlassian.net/wiki/spaces/Technology/pages/734528427/GitHub+-+Gestione+Secrets

> 6. Added as Admin the bot [devops-pagopa-github-bot](https://github.com/devops-pagopa-github-bot) to use the PAT required for modifying repository variables.

Have you tried using ${{ secrets.GITHUB_TOKEN }}? I think is possibile.

File: .github/workflows/check-release.yml
Comment (pasqualedevita) [inline .github/workflows/check-release.yml:74]: if we can't use ${{ secrets.GITHUB_TOKEN }} I suggest to write this information in a file so we reduce permission scopes

Comment (Krusty93) [reply .github/workflows/check-release.yml:74]: Personally, I prefer to not have repository changes through pipelines.
For instance, what does it happen if a PR is not merged in the same day it is opened? I think we would end up having a pull request for each day until one of them is merged, as the workflow always compare the versioned file with what is found on the runner repository.

Using repository/environment variable allows us to update that value when the first PR is opened, avoiding clones.

Unfortunately, we need a PAT as `secrets.GITHUB_TOKEN` is not enough (GitHub permission model is a shame), and I see the bot as the best way to implement it (if you can think of a better solution I'm open). *Unfortunately*, this is not the first time we have to use bots..

**Comment by mamu0** [comment]
Thank you, @pasqualedevita, for your review!
Regarding the first point:

> 4\. Enabling “Allow GitHub Actions to create and approve pull requests” in Repository Settings -> Actions -> General.

This configuration is necessary to allow the GitHub Action to create the PR.
As for this,

> Please check that all new secrets follow the technology standard to prevent leak or exposure.
 
I created the environment and added the secrets to it to adhere to the standard.

**Comment by pasqualedevita** [comment]
> Thank you, @pasqualedevita, for your review! Regarding the first point:
> 
> > 4. Enabling “Allow GitHub Actions to create and approve pull requests” in Repository Settings -> Actions -> General.

done I'm afraid to "approve pull requests" but is ok we have CODEOWERS required

> This configuration is necessary to allow the GitHub Action to create the PR. As for this,
> 
> > Please check that all new secrets follow the technology standard to prevent leak or exposure.
> 
> I created the environment and added the secrets to it to adhere to the standard.

thank you

**Comment by mamu0** [comment]
> done I'm afraid to "approve pull requests" but is ok we have CODEOWERS required

I understand, but unfortunately, that option is not available specifically for creation.

File: .github/workflows/check-release.yml
Comment (mamu0) [reply .github/workflows/check-release.yml:73]: I agree with @Krusty93, with the file as long as it is not merged the workflow will continue to find a release to update. 
But if we want to use a file the only alternative that comes to mind is to do a step that checks the existence of the branch `update-release-to-${{ steps.release_outputs.outputs.tag_name }}` which is normally created if there is an update to be done. 
In this way you can have a control similar to that of environment variables.

**Comment by pasqualedevita** [comment]
> > done I'm afraid to "approve pull requests" but is ok we have CODEOWERS required
> 
> I understand, but unfortunately, that option is not available specifically for creation.

it's ok we have codeowners :) and the flag is enabled. Opened pul request will not be merged automatically

File: .github/workflows/check-release.yml
Comment (pasqualedevita) [reply .github/workflows/check-release.yml:73]: In my review this is the last open point.
I remember that the write access can modify the secrets. Can we try to use secrets instead of variables?

Comment (mamu0) [reply .github/workflows/check-release.yml:73]: It can’t be done due to insufficient permissions granted to the job (also `write-all` is insufficient), but this might be unnecessary. It doesn’t matter if that value is visible to everyone. It’s simply a check to identify the resource currently in use, and if it gets altered, it would only result in a PR that doesn’t need to be merged.

Comment (pasqualedevita) [reply .github/workflows/check-release.yml:73]: So I think that suggested to checks the existence branch with naming convention can be a good tradeoff

`update-release-to-${{ steps.release_outputs.outputs.tag_name }}`

**Comment by Krusty93** [comment]
> for slack notification you can also use github integration with summary. This not realtime but I think that one or two trigger per day can be enough
> 
> Less code is better :)

Does it send an email everyday or only if there is at least a PR?

**Review by Krusty93 — APPROVED** [review]

**Review by pasqualedevita — APPROVED** [review]

**Comment by pasqualedevita** [comment]
thanks @mamu0 very useful improvement :)))

**Pull request merged**
